src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,505 @@
+
+  for (Real v: likelihoods) { 
+   
+    if (v <= redThreshold) { //TODO review the IFs
+      // Value is in the redzone
+        
+      if (filteredLikelihoods.back() > redThreshold) {
+        // Previous value is not in redzone, so leave as-is
+          filteredLikelihoods.push_back(v);
+      } else { 
+        filteredLikelihoods.push_back(yellowThreshold);
+      }
+    } else {
+      // Value is below the redzone, so leave as-is
+      filteredLikelihoods.push_back(v);
+    }
+  }
+  return filteredLikelihoods;
+}
+
+
+vector<Real>  AnomalyLikelihood::updateAnomalyLikelihoods(vector<Real> anomalyScores, UInt verbosity) { 
+  /**
+  Compute updated probabilities for anomalyScores using the given params.
+
+  :param anomalyScores: a list of records. Each record is a list with the
+                        following three elements: [timestamp, value, score]
+
+                        Example::
+
+                            [datetime.datetime(2013, 8, 10, 23, 0), 6.0, 1.0]
+
+  :param params: the JSON dict returned by estimateAnomalyLikelihoods
+  :param verbosity: integer controlling extent of printouts for debugging
+  :type verbosity: int
+
+  :returns: 3-tuple consisting of:
+
+            - likelihoods
+
+              numpy array of likelihoods, one for each aggregated point
+
+            - avgRecordList 
+
+              list of averaged input records - stored in runningAverageAnomalies instance member
+
+            - params
+
+              an updated JSON object containing the state of this metric.-- available from historicalScores.getXXX()
+
+  **/
+  if (verbosity > 3) {
+    cout << "In updateAnomalyLikelihoods."<< endl;
+    cout << "Number of anomaly scores: "<<  anomalyScores.size() << endl;
+//    cout << "First 20:", anomalyScores[0:min(20, len(anomalyScores))])
+    cout << "Params: name=" <<  distribution.name << " mean="<<distribution.mean <<" var="<<distribution.variance <<" stdev="<<distribution.stdev <<endl;
+  }
+
+ assert(anomalyScores.size() > 0); // "Must have at least one anomalyScore"
+
+  // Compute moving averages of these new scores using the previous values
+  // as well as likelihood for these scores using the old estimator 
+  vector<Real> likelihoods;
+  for (auto newAverage : runningAverageAnomalies) {
+    likelihoods.push_back(tailProbability(newAverage)); 
+  }
+
+  // Filter the likelihood values. First we prepend the historical likelihoods
+  // to the current set. Then we filter the values.  We peel off the likelihoods
+  // to return and the last windowSize values to store for later.
+  UInt toCrop = min((unsigned long)this->averagedAnomaly.getMaxSize(), (unsigned long)runningLikelihoods.size());
+  this->runningLikelihoods.insert(runningLikelihoods.end() - toCrop, likelihoods.begin(),likelihoods.end()); //append & crop
+  assert(this->runningLikelihoods.size() <= this->averagedAnomaly.getMaxSize());
+
+  auto filteredLikelihoods = filterLikelihoods_(likelihoods);
```

Comment:
Isn't this another use-case for `auto`? It allows us to change the return-type in the method easity (as long as it's compatible)...